### DISCUSSION TITLE
Async profiler unit tests

### DISCUSSION THREAD

**Discussion by toktarev** [async-profiler/async-profiler#495] [Ideas]
This is continuation of the previous discussion: https://github.com/jvm-profiling-tools/async-profiler/issues/493

**Comment by toktarev** [comment]
I think it would be good code quality improvement if we would cover  by fine-grained unit tests all internal profiler's abstractions
For example set of self-written hash tables.

**Reply by apangin** [reply]
Pardon, I'm not sure I get this corectly - who do you mean by "we" here? Are you suggesting me to do this, or do you offer to add this yourself?

**Comment by toktarev** [comment]
Saying "we" I mean people from the community.

Of course, I can do it, if you don't have any objections.

But maybe you don't want me to do this and someone else could do this.

This is why I used "we".

**Reply by apangin** [reply]
To be honest, I don't want you, or myself, or someone else to do this, because the proposed idea does not solve any real existing problem. E.g. the allocator has been written just once, tested, integrated, stress-tested on real applications, and now used in production, and there are no plans to touch it in the nearest future. Adding unit tests for it *now* will likely cause more issues than it solves.

On the other hand, functional tests make a lot more sense - to make sure the expected behavior of async-profiler (from user's perspective) does not change from release to release, from one JDK version to another, from Linux to macOS, etc. That's what worth investing in, and that's what I'm working on now. If you really want to contribute to the project, it's probably the best thing I'd suggest to start with.

**Comment by toktarev** [comment]
----the proposed idea does not solve any real existing problem.

From my point of view (and as I see in many open source projects) complicated code should be covered by such tests to provide the codebase's stability in case of changes in this code.

For example, LonHashTable refactoring will be protected by such tests.

But if you don't consider such things as something useful - ok I will not insist.

I suggest to close this discussion.

---If you really want to contribute to the project

I don't write here something to find an approval for commits to the async profiler.
Just trying to learn more about the product - because I use it quite often in real systems.
